### PR TITLE
cleanup(scheduler): move metricRecorder to metrics package

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

https://github.com/kubernetes/kubernetes/issues/115083 wants to record `plugin_execution_duration_seconds` metric for PreEnqueue in the scheduling queue. 
Currently, all extension points record `plugin_execution_duration_seconds` via [`metricsRecorder`](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/metrics_recorder.go#L29) which records the metrics in a separate goroutine. Then, it's better to utilize this for PreEnqueue as well for better performance.

The scheduling queue package ("k8s.io/kubernetes/pkg/scheduler/internal/queue") is outside of the framework package and the `metricsRecorder` in the [frameworkImpl](https://github.com/kubernetes/kubernetes/blob/0e3818e02760afa8ed0bea74c6973f605ca4683c/pkg/scheduler/framework/runtime/framework.go#L96) struct isn't callable from the scheduling queue.

This PR just moves it to `pkg/scheduler/metrics/metric_recorder.go` with renaming it to `MetricsAsyncRecorder` so that we can initialize `MetricsAsyncRecorder` outside the framework and use the same `MetricsAsyncRecorder` object both in the framework and the scheduling queue (will be done in another PR).

#### Which issue(s) this PR fixes:

Related to https://github.com/kubernetes/kubernetes/issues/115083

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/metrics/metric_recorder.go [hunk 1]
```diff
@@ -77,3 +79,79 @@ func (r *PendingPodsRecorder) Dec() {
...
+		stopCh:      stopCh,
+		IsStoppedCh: make(chan struct{}),
+	}
+	go recorder.run()
+	return recorder
+}
+
+// ObservePluginDurationAsync observes the plugin_execution_duration_seconds metric.
+// The metric will be flushed to Prometheus asynchronously.
+func (r *MetricAsyncRecorder) ObservePluginDurationAsync(extensionPoint, pluginName, status string, value float64) {
```
Comment (Huang-Wei) [inline pkg/scheduler/metrics/metric_recorder.go:121]: is it intentional to change the type of status from Status to string?

Comment (sanposhiho) [reply pkg/scheduler/metrics/metric_recorder.go:121]: Yes. We cannot use `k8s.io/kubernetes/pkg/scheduler/framework` here due to the import cycle.

```
# When we import k8s.io/kubernetes/pkg/scheduler/framework here,
# and build the scheduler.

package k8s.io/kubernetes/pkg/scheduler/metrics
	imports k8s.io/kubernetes/pkg/scheduler/framework
	imports k8s.io/kubernetes/pkg/scheduler/framework/parallelize
	imports k8s.io/kubernetes/pkg/scheduler/metrics: import cycle not allowed
```

File: pkg/scheduler/metrics/metric_recorder.go [hunk 2]
```diff
@@ -17,7 +17,9 @@ limitations under the License.
 package metrics
 
 import (
-	"k8s.io/component-base/metrics"
+	"time"
+
+	k8smetrics "k8s.io/component-base/metrics"
```
Comment (kerthcet) [inline pkg/scheduler/metrics/metric_recorder.go:22]: /hold
can we use metrics directly here?

Comment (sanposhiho) [reply pkg/scheduler/metrics/metric_recorder.go]: Good catch. Fixed.

**Comment by kerthcet** [comment]
What do you mean `use the same MetricsAsyncRecorder object both in the framework and the scheduling queue`, they're different recorder. @sanposhiho

**Comment by sanposhiho** [comment]
I'm not sure I understood correctly what you mean but I'd like to do (in the next PR) is:
- store the same MetricsAsyncRecorder object in [PriorityQueue](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/internal/queue/scheduling_queue.go#L143) struct.
- run `priorityqueue.metricsRecorder.ObservePluginDurationAsync("PreEnqueue", pl.Name(), status.Code().String(), metrics.SinceInSeconds(startTime))` in [`runPreEnqueuePlugins`](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/internal/queue/scheduling_queue.go#L322).

**Comment by sanposhiho** [comment]
Why the "same" MetricsAsyncRecorder object? -> Each `MetricsAsyncRecorder` object has one goroutine to flush metrics periodically. To reduce the num of goroutines working in the scheduler, it's better to use the same MetricsAsyncRecorder object in both the framework and the priorityqueue.

**Comment by kerthcet** [comment]
Got the idea, I thought the `same record` was https://github.com/kubernetes/kubernetes/blob/1e84987baccbccf929eba98af1d8b19b120e39f3/pkg/scheduler/internal/queue/scheduling_queue.go#L298

/lgtm
/hold cancel
/retest

**Pull request merged**